runtime/ops/pooling.cc

```diff
@@ -135,6 +135,132 @@ chainerx::Array ROIPool2D(const chainerx::Array& bottom_data, const chainerx::Ar
+        int64_t a_high;
+        if(a_low >= len - 1) {
+            a_high = a_low = len - 1;
+            a = static_cast<double>(a_low);
+        }
+        else {
+            a_high = a_low + 1;
+        }
+        double la = a - a_low;
+        double ha = 1.0 - la;
+        return std::make_tuple(a_low, a_high, la, ha);
+    };
+    int64_t y_low, y_high;
+    double ly, hy;
+    std::tie(y_low, y_high, ly, hy) = calc_a_low_a_high_la_ha(y, height);
+    int64_t x_low, x_high;
+    double lx, hx;
+    std::tie(x_low, x_high, lx, hx) = calc_a_low_a_high_la_ha(x, width);
+    double w1 = hy * hx;
+    double w2 = hy * lx;
+    double w3 = ly * hx;
+    double w4 = ly * lx;
+    return std::make_tuple(y_low, x_low, y_high, x_high, w1, w2, w3, w4);
+}
+
+template<class ReduceMode>
+chainerx::Array ROIAlign2D(const chainerx::Array& bottom_data, const chainerx::Array& bottom_rois, const chainerx::Array& bottom_roi_indices, const Int64StackVector& output_shape, const float spatial_scale, const chainerx::StackVector<int64_t, chainerx::kMaxNdim>& sampling_ratio) {
+    CHECK_EQ(4, bottom_data.ndim());
+    CHECK_EQ(2, output_shape.size());
+    CHECK_EQ(2, sampling_ratio.size());
+
+    const int64_t channels = bottom_data.shape()[1];
+    const int64_t height = bottom_data.shape()[2];
+    const int64_t width = bottom_data.shape()[3];
+    const int64_t n_rois = bottom_rois.shape()[0];
+    const int64_t pooled_height = output_shape[0];
+    const int64_t pooled_width = output_shape[1];
+    chainerx::Array top_data = chainerx::Zeros(chainerx::Shape{n_rois, channels, pooled_height, pooled_width}, bottom_rois.dtype());
+
+    for(int64_t n = 0; n < n_rois; ++n) {
+        int64_t roi_batch_ind = int64_t(chainerx::AsScalar(bottom_roi_indices.At({n})));
+        double roi_start_h = double(chainerx::AsScalar(bottom_rois.At({n, 0})) * spatial_scale);
+        double roi_start_w = double(chainerx::AsScalar(bottom_rois.At({n, 1})) * spatial_scale);
+        double roi_end_h = double(chainerx::AsScalar(bottom_rois.At({n, 2})) * spatial_scale);
+        double roi_end_w = double(chainerx::AsScalar(bottom_rois.At({n, 3})) * spatial_scale);
+        double roi_height = std::max<double>(roi_end_h - roi_start_h, 1.);
+        double roi_width = std::max<double>(roi_end_w - roi_start_w, 1.);
+        double bin_size_h = roi_height / pooled_height;
+        double bin_size_w = roi_width / pooled_width;
+
+        int64_t roi_bin_grid_h = sampling_ratio[0];
+        int64_t roi_bin_grid_w = sampling_ratio[1];
+
+        for(int64_t c = 0; c < channels; ++c) {
+            for (int64_t ph = 0; ph < pooled_height; ++ph) {
+                for (int64_t pw = 0; pw < pooled_width; ++pw) {
+                    ReduceMode reduce;
+                    int64_t iy = 0;
+                    while(iy < roi_bin_grid_h) {
+                        auto y = roi_start_h + ph * bin_size_h + (iy + 0.5) * bin_size_h / roi_bin_grid_h;
```

Comment:
I see. Personally, I use `auto` anytime it can be used because it allows quick modification of sources especially about types. However, the types used in codes here are almost fixed so readability is more important. I adopt your method.